src/client/client.ts

```diff
@@ -42,6 +45,10 @@ type ClientAction =
   | ActionShape.Patch;
 type Action = CredentialedActionShape.Any | ClientAction;
 
+// TODO(#723): Figure out a nice API for the returned ActionResult.
```

Comment:
@delucis: This is basically the crux to the issues you raised:

1. TransientHandlingMiddleware is meant to be used (eventually) by all invokers of the middleware. I wanted it to have some common interface, which for the sake of discussion, let's call `ActionWithTransients`. It should have exactly the type that you described:
    ```typescript
    type MiddlewareReturn = ActionShape.Any & {
      stripTransientsResult?: ActionShape.StripTransients;
      transients?: TransientMetadata;
    };
    ```

2. At this point the handling on the client and master diverge somewhat but both can use this `ActionWithTransients` to deal with their concerns either with divergent middlewares sitting in front of the `TransientsHandlingMiddleware` or by directly wrapping the return value of the dispatch calls. For example, the master's handling can pretty much just use `createErrorHandlingMiddleware` slightly modified as follows:

    ```typescript
    const createResultHandlingMiddleware =
      (onSuccess: (actionResult: SomeGenericSuccessfulReturnType) => void) =>
      (onError: (error: ActionError) => void) =>
      (store: Store) =>
      (next: Dispatch<ActionShape.Any>) =>
      (action: ActionShape.Any) => {
        // Delegate to the stripping handler.
        const result = next(action);
        if (action.type instanceof VirtualActions /*i.e. Actions.STRIP_TRANSIENTS */) {
           // Trust the underlying middleware to handle these "virtual" actions.
           return result;
        }
        // At this point we should just handle remaining actions as we would.
        assert ! action.type instanceof VirtualActions;
        const { transients } = result;
        if (typeof transients !== 'undefined') {
          // Leave re-dispatch and result multi-plexing to the TransientHandlingMiddleware.
          if (transients.error) onError(transients.error);
        }
        return onSuccess(new SomeGenericSuccessfulReturnType(result));
      };
    ```

3. The only reason why I'm not super sure that pinning the above change and the typing of ActionResult here to the current change is ideal is because I think the most reasonable type would be a breaking change:

    ```typescript
    /* The promise will not resolve until the reducer has actually processed the action!!! */
    export type ActionResult = Promise<NiceGeneralResultBody>;
    export type NiceGeneralResultBody = any /* The actual data model of action results */;
    ```
 
Returning this promise with this semantic, for example requires all callers (even for external users of the API) to _have_ to `await` the result of their `move.Foo()` calls in order to get the correct behavior. I would probably want to work around this by leaving existing synchronous APIs that directly return `NiceGeneralResultBody`.

In my naive view, biting off this next piece requires a significant ballooning of scope in order to restore backwards compatibility. In the current state, the backwards-incompatibility is resolved with the argument that the return value isn't nice and user-consumable, but the previous return value was Void and there's at least a type indicator that we can fix up (and document) once we're ready to export it to users.